### DISCUSSION TITLE
Nonebot Plugin Alconna 0.46.0

### DISCUSSION THREAD

**Discussion by RF-Tar-Railt** [nonebot/plugin-alconna#50] [Announcements]
## 新增
- 适配器 [Mirai(官方)](https://github.com/nonebot/adapter-mirai)
- 适配器 [TailChat](https://github.com/eya46/nonebot-adapter-tailchat)
- `Text` 新增 `.bold()`, `.italic()` 等一系列便捷方法
- 新增 `I18n` 通用消息段（在发送前会转为 UniMessage）
- `AlconnaMatcher` 新增 `.i18n()` 方法，与  `I18n` 作用相同
- 新增内置插件 `lang`，可以切换或列出可用的语言模式

## 改进
- 升级 `Tarina` 至 0.5.0 以使用 tarina-lang 新特性

## i18n 说明

plugin-alconna 的 i18n 支持基于 [`tarina.lang`](https://github.com/ArcletProject/Tarina/tree/main/src/tarina/lang)，其提供了一个 tarina-lang 命令行工具

首先可以通过 `tarina-lang new` 创建文件夹 `i18n`

之后使用 `cd ./i18n` 和 `tarina-lang init`，会生成如下文件：
```diff
📦 awesome-bot
├──📂 i18n
++ ├── __init__.py
++ ├── .config.json
++ ├── .template.json
++ └── .template.schema.json
├── xxx.py
└── ...
```

你需要将你语言文件中所有包含的项目声明在 `.template.json` 中，例如：

```json
{
  "$schema": ".template.schema.json",
  "scopes" : [
    {
      "scope": "example",
      "types": [
        "test",
        {
          "subtype": "test1",
          "types": [
            "test2"
          ]
        }
      ]
    }
  ]
}
```

然后通过 `tarina-lang schema` 和 `tarina-lang create XXX` 来创建新的语言文件。以下为使用命令创建 `en-US` 和 `zh-CN` 语言文件后的文件结构：
```diff
📦 awesome-bot
├──📂 i18n
│  ├── __init__.py
│  ├── .config.json
++ ├── .lang.schema.json
│  ├── .template.json
│  ├── .template.schema.json
++ ├── en-US.json
++ └── zh-CN.json
├── plugin.py
└── ...
```

其中一个语言文件如下所示：

```json5
// en-US.json
{
  "$schema": "./.lang.schema.json",
  "example": {
    "test": "Test",
    "test1": {
      "test2": "Test2"
    }
  }
}
```

> [!NOTE]
> `tarina-lang` 支持创建和读取 YAML 格式的语言文件。当然首先你需要额外安装 `tarina[yaml]`
> 
> 然后通过 `tarina-lang create XXX --yaml` 创建 `.yml` 文件
>
> 一个 yaml 格式的语言文件如下所示：
```yaml
# $schema: .lang.schema.json
example:
  test: Test
  test1:
    test2: Test2
```

之后，在 `plugin` 里面，你可以用如下方法来使用i18n条目：

```python
from .i18n import lang

...
async def _():
    await matcher.send(lang.require("example", "test"))  # Test
    await matcher.send(lang.require("example", "test1.test2"))  # Test2
```

高级一点，你可以通过 `tarina-lang model` 来生成一个模型文件：

```diff
📦 awesome-bot
├──📂 i18n
│  ├── __init__.py
│  ├── .config.json
│  ├── .lang.schema.json
│  ├── .template.json
│  ├── .template.schema.json
│  ├── en-US.json
++ ├── model.py
│  └── zh-CN.json
├── plugin.py
└── ...
```

其中 `model.py`:

```python
from tarina.lang.model import LangItem, LangModel


class ExampleTest1:
    test2: LangItem = LangItem("example", "test1.test2")


class Example:
    test: LangItem = LangItem("example", "test")
    test1: ExampleTest1


class Lang(LangModel):
    example = Example

```

之后便可以这样使用：

```python
from .i18n import Lang

...
async def _():
    await matcher.send(Lang.example.test())
    # 如果你的条目是模板字符串，你可以使用 Lang.example.test(...)
    await matcher.send(Lang.example.test1.test2())
```

基于此，`I18n` 和 `AlconnaMatcher.i18n` 可以如下使用：

```python
await AlconnaMatcher.send(UniMessage.i18n(Lang.example.test, ...))
await AlconnaMatcher.send(UniMessage.i18n("example", "test", ...))
await AlconnaMatcher.send(AlconnaMatcher.i18n(Lang.example.test1.test2, ...))
await AlconnaMatcher.send(AlconnaMatcher.i18n("example", "test1.test2", ...))
```

> [!NOTE]
> 在 `plugin-alconna` 中, i18n 条目会先被转换成 UniMessageTemplate
> 所以 UniMessageTemplate 的所有特性都可用于 i18n 条目
> 例如：
```python
# example.test: "{:At(user, $event.get_user_id())} Hello!"
await XXX.send(XXX.i18n("example", "test"))
```
> 特别的，因为 `I18n` 是一个通用消息段，所以 i18n 条目可以嵌套：
```json
{
  "example": {
    "test": "XXXX",
    "foo": "{:I18n(example, test)}, XXXX!"
  }
}
```

**Full Changelog**: https://github.com/nonebot/plugin-alconna/compare/v0.45.4...v0.46.0

<hr /><em>This discussion was created from the release <a href='https://github.com/nonebot/plugin-alconna/releases/tag/v0.46.0'>Nonebot Plugin Alconna 0.46.0</a>.</em>

**Comment by Agnes4m** [comment]
看不懂喵，有没有使用的插件案例或者demo，

**Reply by RF-Tar-Railt** [reply]
nonebot-plugin-inspect

**Reply by Agnes4m** [reply]
那切换语言的指令是什么

**Reply by RF-Tar-Railt** [reply]
> 那切换语言的指令是什么

lang switch XXX

**Reply by Agnes4m** [reply]
![image](https://github.com/user-attachments/assets/4c37b71f-608e-4886-b236-49eeeea0b8fa)
mei'y没有响应捏

**Reply by RF-Tar-Railt** [reply]
> ![image](https://github.com/user-attachments/assets/4c37b71f-608e-4886-b236-49eeeea0b8fa)
> mei'y没有响应捏

你启用内置插件了吗（

https://nonebot.dev/docs/next/best-practice/alconna/builtins#%E5%86%85%E7%BD%AE%E6%8F%92%E4%BB%B6

**Reply by Agnes4m** [reply]
我这么启用的应该mei'c没错吧
![image](https://github.com/user-attachments/assets/b8335a60-6844-42b3-a786-a5c053059912)

**Reply by RF-Tar-Railt** [reply]
> 我这么启用的应该mei'c没错吧
> ![image](https://github.com/user-attachments/assets/b8335a60-6844-42b3-a786-a5c053059912)
> 

你看你启动的时候有没有加载的log，然后检查一下你的command_start

**Reply by Agnes4m** [reply]
启动了确定没问题的，感觉就是指令不对
![image](https://github.com/user-attachments/assets/59bb34b1-0bbf-4d6d-b293-e39e6ad6a8ef)

![image](https://github.com/user-attachments/assets/4d94386a-5bf8-4cf8-947e-98795add4a40)

**Reply by RF-Tar-Railt** [reply]
日志等级开trace看看

**Reply by Agnes4m** [reply]
![image](https://github.com/user-attachments/assets/a2542717-f7d9-4fd4-b0bb-9f9ab4409d3a)

**Reply by RF-Tar-Railt** [reply]
，，，，是 switch 不是 swich

**Reply by Agnes4m** [reply]
ok，没问题了，是我输入错误的原因

**Comment by Agnes4m** [comment]
使用model方式后，是否只需要修改model.py而可以无视.lang.schema.json的修改

**Reply by RF-Tar-Railt** [reply]
model和schema的更新只有template修改后才需要

**Reply by Agnes4m** [reply]
意思就是，如果我需要新增或者删除字段，只需要修改template，然后再自动生成model或者schema，对吗

**Reply by RF-Tar-Railt** [reply]
> 意思就是，如果我需要新增或者删除字段，只需要修改template，然后再自动生成model或者schema，对吗

对

**Comment by Agnes4m** [comment]
如何在文件中使用i18n翻译后的文字？
我在使用jinja渲染时遇到问题，
- LangItem对象并不能作为str输出，因此jinja模板后，html里会变成类似`<tarina.lang.model.LangItem object at 0x000001CB67712AE0>`的值
- __str__()也一样返回类似上面的值而不是文字
- 字段`scope`和`type`只能返回上层分类的值
- 以__call__()方式会报错如下
```
    print(Lang.nbp_vrc.words.online.__call__())
  File "d:\截图\其他文件\workspace\.venv\Lib\site-packages\tarina\lang\model.py", line 20, in __call__
    return lang.require(self.scope, self.type, locale).format(**format_kwargs)
  File "d:\截图\其他文件\workspace\.venv\Lib\site-packages\tarina\lang\__init__.py", line 304, in require
    raise ValueError(self.__langs[locale]["lang"]["error.type"].format(target=type, locale=locale, scope=scope))
ValueError: 'words.online' 在 'zh-CN:nbp_vrc' 不是合法的类型
```
![image](https://github.com/user-attachments/assets/7c032343-534d-4668-b5d2-dbdcecd09e1f)

**Reply by RF-Tar-Railt** [reply]
LangItem 就是直接像 `Lang.nbp_vrc.words.online()` 这样用的

至于这个报错得看你语言文件了

**Reply by Agnes4m** [reply]
确实是我语言文件写错了，不好意思
